File: samples-java/src/test/java/io/github/ilaborie/loops/samples/NestTest.java
package io.github.ilaborie.loops.samples;

import org.junit.Test;

import java.util.Arrays;
import java.util.List;
import java.util.function.Function;

import static org.junit.Assert.assertEquals;

public class NestTest {


    private void test(Function<List<Element>, List<String>> func) {
        List<Element> elements = Arrays.asList(
                new Element("a"),
                new Element("b"),
                new Element("aa"));

        List<String> result = func.apply(elements);

        assertEquals(Arrays.asList("a-b", "b-a"), result);
    }

    @Test
    public void for_should_nest() {
        test(io.github.ilaborie.loops.samples.anatomy.Nest::nest);
    }

    @Test
    public void stream_should_nest() {
        test(io.github.ilaborie.loops.samples.stream.Nest::nest);
    }

    @Test
    public void stream_should_nest_alt() {
        test(io.github.ilaborie.loops.samples.stream.Nest::nestAlt);
    }
}

File: samples-kotlin/src/test/kotlin/io/github/ilaborie/loops/samples/stream/NestTest.kt
package io.github.ilaborie.loops.samples.stream

import io.github.ilaborie.loops.samples.ElementKt
import org.junit.Assert.assertEquals
import org.junit.Test

class NestTest {

    private fun test(func: (List<ElementKt>) -> List<String>) {
        val elements = listOf(
            ElementKt("a"),
            ElementKt("b"),
            ElementKt("aa")
        )

        val result = func(elements)

        assertEquals(listOf("a-b", "b-a"), result)
    }


    @Test
    fun for_should_nest() {
        test { io.github.ilaborie.loops.samples.anatomy.nest(it) }
    }

    @Test
    fun stream_should_nest() {
        test { io.github.ilaborie.loops.samples.stream.nest(it) }
    }

    @Test
    fun stream_should_nest_seq() {
        test { io.github.ilaborie.loops.samples.stream.nestSeq(it) }
    }
}
